ext/pdo/pdo_stmt.c

```diff
@@ -1216,37 +1163,34 @@ PHP_METHOD(PDOStatement, fetchAll)
 	int flags = how & PDO_FETCH_FLAGS;
 
 	old_ce = stmt->fetch.cls.ce;
-	ZVAL_COPY_VALUE(&old_ctor_args, &stmt->fetch.cls.ctor_args);
-	old_arg_count = stmt->fetch.cls.fci.param_count;
-
-	do_fetch_opt_finish(stmt, 0);
+	old_ctor_args = stmt->fetch.cls.ctor_args;
 
 	/* TODO Would be good to reuse part of pdo_stmt_setup_fetch_mode() in some way */
-
 	switch (fetch_mode) {
 		case PDO_FETCH_CLASS:
 			/* Figure out correct class */
 			if (arg2) {
 				if (Z_TYPE_P(arg2) != IS_STRING) {
 					zend_argument_type_error(2, "must be of type string, %s given", zend_zval_value_name(arg2));
-					RETURN_THROWS();
+					goto reset_old_ctor_args;
 				}
 				stmt->fetch.cls.ce = zend_lookup_class(Z_STR_P(arg2));
 				if (!stmt->fetch.cls.ce) {
 					zend_argument_type_error(2, "must be a valid class");
-					RETURN_THROWS();
+					goto reset_old_ctor_args;
```

Comment:
Now you'll jump to `reset_old_ctor_args` where you also release the `ctor_args` array, but this was not set already at this point. So this can lead to a refcount bug.
What I would do instead is store the result of `zend_lookup_class` in a variable, and only assign the result of that variable to `stmt->fetch.cls.ce` _after_ the NULL check on that variable. This allows you to keep using `RETURN_THROWS();` and not reason too much about the restoring.

Reply:
Took this idea even if no need to refcount